toolchain/check/merge.cpp

```diff
@@ -232,78 +229,101 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
 
-  if (new_param_pattern.Is<SemIR::AnyParamPattern>()) {
-    new_param_pattern_id =
-        new_param_pattern.As<SemIR::ValueParamPattern>().subpattern_id;
-    new_param_pattern = context.insts().Get(new_param_pattern_id);
-    prev_param_pattern_id =
-        prev_param_pattern.As<SemIR::ValueParamPattern>().subpattern_id;
-    prev_param_pattern = context.insts().Get(prev_param_pattern_id);
+  pattern_stack.push_back({.prev_id = orig_prev_param_pattern_id,
+                           .new_id = orig_new_param_pattern_id});
+
+  do {
+    auto patterns = pattern_stack.pop_back_val();
+    auto new_param_pattern = context.insts().Get(patterns.new_id);
+    auto prev_param_pattern = context.insts().Get(patterns.prev_id);
     if (new_param_pattern.kind() != prev_param_pattern.kind()) {
       emit_diagnostic();
       return false;
     }
-  }
 
-  auto new_name_id =
-      context.entity_names()
-          .Get(new_param_pattern.As<SemIR::AnyBindingPattern>().entity_name_id)
-          .name_id;
-  auto prev_name_id =
-      context.entity_names()
-          .Get(prev_param_pattern.As<SemIR::AnyBindingPattern>().entity_name_id)
-          .name_id;
-
-  if (!check_self && new_name_id == SemIR::NameId::SelfValue &&
-      prev_name_id == SemIR::NameId::SelfValue) {
-    return true;
-  }
+    switch (new_param_pattern.kind()) {
+      case SemIR::AddrPattern::Kind: {
+        pattern_stack.push_back(
+            {.prev_id = prev_param_pattern.As<SemIR::AddrPattern>().inner_id,
+             .new_id = new_param_pattern.As<SemIR::AddrPattern>().inner_id});
+        break;
+      }
+      case SemIR::OutParamPattern::Kind:
+      case SemIR::RefParamPattern::Kind:
+      case SemIR::ValueParamPattern::Kind: {
+        pattern_stack.push_back(
+            {.prev_id =
+                 prev_param_pattern.As<SemIR::AnyParamPattern>().subpattern_id,
+             .new_id =
+                 new_param_pattern.As<SemIR::AnyParamPattern>().subpattern_id});
+        break;
+      }
+      case SemIR::VarPattern::Kind:
+        pattern_stack.push_back(
+            {.prev_id =
+                 prev_param_pattern.As<SemIR::VarPattern>().subpattern_id,
+             .new_id =
+                 new_param_pattern.As<SemIR::VarPattern>().subpattern_id});
+        break;
+      case SemIR::BindingPattern::Kind:
+      case SemIR::SymbolicBindingPattern::Kind: {
+        auto new_name_id =
+            context.entity_names()
+                .Get(new_param_pattern.As<SemIR::AnyBindingPattern>()
+                         .entity_name_id)
+                .name_id;
+        auto prev_name_id =
+            context.entity_names()
+                .Get(prev_param_pattern.As<SemIR::AnyBindingPattern>()
+                         .entity_name_id)
+                .name_id;
+
+        if (!check_self && new_name_id == SemIR::NameId::SelfValue &&
+            prev_name_id == SemIR::NameId::SelfValue) {
+          return true;
```

Comment:
Seems like this ignores other elements on the stack. Or is this relying on there never being more than a single item on the `pattern_stack`? That seems undocumented and likely fragile.

```suggestion
          break;
```

Reply:
I think the idea is that if `check_self` is `false`, we don't want to do any checking of the pattern that defines the `self` parameter. That said, are we returning early enough in that case? In particular, it looks to me like for `addr self: Self*`, we'll compare the type of the `AddrPattern` before we reach the self pattern and would diagnose if a redeclaration has a different `self`. How does this work?

Reply:
> Seems like this ignores other elements on the stack. Or is this relying on there never being more than a single item on the `pattern_stack`? That seems undocumented and likely fragile.

Good catch; fixed.

> I think the idea is that if `check_self` is `false`, we don't want to do any checking of the pattern that defines the `self` parameter. That said, are we returning early enough in that case? In particular, it looks to me like for `addr self: Self*`, we'll compare the type of the `AddrPattern` before we reach the self pattern and would diagnose if a redeclaration has a different `self`. How does this work?

Based on what breaks if I remove this special case, the intent here seems to be to work around the problem that, when declaring an override of a base class method, a syntactically identical redeclaration would fail, because `Self` names a different type than it did in the base class. So the goal isn't to bypass all checking of the pattern, it's to bypass checking that the types match. That happens only at the binding pattern, not at e.g. the `addr` pattern, presumably because the type of a non-leaf pattern is always determined by the types of its children, so it's sufficient to check that the leaf types match and all inst kinds match.